src/main.rs

```diff
@@ -655,86 +691,164 @@ pub fn construct_haplotypes(fa: &mut fasta::IndexedReader<File>,
+            let mut parsed_scores = HashMap::new();
+            for score in cell_scores.into_iter() {
+                let eval = evaluate_scores(score.ref_score, score.alt_score);
+                if eval.is_none() {
+                    continue;
+                }
+                parsed_scores.entry(&score.umi).or_insert(Vec::new()).push(eval.unwrap());
+            }
+            // collapse each UMI into a consensus value
+            let mut collapsed_scores = Vec::new();
+            for (_umi, v) in parsed_scores.into_iter() {
+                let counts = convert_to_counts(v);
+                let ref_frac = counts.ref_count as f64 / (counts.alt_count as f64 + counts.ref_count as f64);
+                let alt_frac = counts.alt_count as f64 / (counts.alt_count as f64 + counts.ref_count as f64);
+                if (counts.unk_count > 1) | ((ref_frac < CONSENSUS_THRESHOLD) & (alt_frac < CONSENSUS_THRESHOLD)) { 
+                    collapsed_scores.push(UNKNOWN_VALUE);
+                }
+                else if alt_frac >= CONSENSUS_THRESHOLD {
+                    collapsed_scores.push(ALT_VALUE);
+                }
+                else {
+                    assert!(ref_frac >= CONSENSUS_THRESHOLD);
+                    collapsed_scores.push(REF_VALUE);
+                }
+            }
+            let counts = CellCalls {cell_index: cell_index, calls: collapsed_scores};
+            r.push(counts);
+        }
+        else {
+            let mut scores = Vec::new();
+            for score in cell_scores.into_iter() {
+                let _eval = evaluate_scores(score.ref_score, score.alt_score);
+                if _eval.is_none() {
+                    continue;
+                }
+                let eval = _eval.unwrap();
+                scores.push(eval);
+            }
+            let counts = CellCalls {cell_index: cell_index, calls: scores};
+            // map of CB to Score objects. This is basically trivial in the non-UMI case
+            r.push(counts);
+        }
+    }
+    r
+}
+
+
+pub fn consensus_scoring(results: &EvaluateAlnResults, i: usize, umi: bool) -> Vec<(u32, f64)> {
+    let parsed_scores = parse_scores(&results.scores, umi);
     let mut result = Vec::new();
-    for (bc, r) in parsed_scores.into_iter() {
-        let counts = convert_to_counts(r);
+    for s in parsed_scores.into_iter() {
+        let counts = convert_to_counts(s.calls);
         if counts.unk_count > 1 {
-            info!("Variant at index {} has multiple unknown reads at barcode index {}. Check this locus manually", i, bc);
+            info!("Variant at index {} has multiple unknown reads at barcode index {}. Check this locus manually", i, s.cell_index);
         }
         
         if (counts.ref_count > 0) & (counts.alt_count > 0) {
-            result.push((bc, REF_ALT_VALUE as f64));
+            result.push((s.cell_index, REF_ALT_VALUE as f64));
```

Comment:
Will do.

Reply:
I am thinking more about this, and I am not sure I understand how this would be better. These three constants are used all over the place, and it seems pretty clear to me. Putting these in an enum will require me to also implement a comparison function for the enum, otherwise the `convert_to_counts` function won't work anymore.

Reply:
You just put 
`#[derive(PartialEq, Eq)]` on top of the type and it will implement comparison for you.

Reply:
Yeah, it's pretty reasonable as it is, this is more of a Rust style suggestion.  Making it it's own type will prevent you from accidentally using the i8 value in a way you didn't intend to.  Probably not critical in this case, but it's a good habit to get into.